cirq/optimizers/controlled_gate_decomposer.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+from cirq import linalg, ops
+import numpy as np
+
+
+def _unitary_power(U, p):
+    """Raises unitary matrix U to power p."""
+    assert linalg.is_unitary(U)
+    eig_vals, eig_vectors = np.linalg.eig(U)
+    Q = np.array(eig_vectors)
+    return Q @ np.diag(np.exp(p * 1j * np.angle(eig_vals))) @ Q.conj().T
+
+
+def _is_identity(M):
+    """Checks whether M is identity."""
+    return np.allclose(M, np.eye(M.shape[0]))
+
+
+def _ry(x):
```

Comment:
`protocols.unitary(cirq.Ry(x))`

Don't duplicate the definition of these matrices.

Reply:
Done.
Btw, there is difference in sign between definition of Ry, Rz in paper and implementation in Cirq.

Reply:
The convention we use in Cirq is the one from Mike and Ike, so we'll stick with it.